rfcs/0128-redeployable-clients.md

```diff
@@ -0,0 +1,393 @@
+```
+
+This is important because not all code will want to configure settings based on
+environment variables, for example workers that take their configuration from
+configuration files.
+
+## Changes to adding taskcluster client as a service dependency
+
+When upgrading to the new client, care will need to be taken to ensure that
+taskcluster root URL is passed into the client. In addition, the
+references/schemas that the client reads dynamically should also be a
+dependency of the project, which may be either frozen references/schemas,
+fetched from a language package, or fetched dynamically during build/CI from a
+taskcluster deployment.
+
+## Changes to projects that depend on a taskcluster client
+
+Those that depend on a client with generated code based on the content of the
+API references and schemas, should either run the code generation tool inside
+their project on an ad hoc basis, and check in the generated code, or should
+run it as part of their build/ci procedure.
+
+All code that uses any of the new clients will need to explicitly pass in a
+root URL in a constructor, or explicitly call a method to fetch
+`TASKCLUSTER_ROOT_URL` from the environment.
+
+## Changes to `taskcluster-proxy` and its configuration
+
+The taskcluster proxy should require an additional property to start up, which
+is the root url of the deployment it should talk to. This should be a mandatory
+property, and no default should be supplied, so the user is forced to make an
+active choice about which environment they wish the proxy to connect to.
+
+## Changes to workers
+
+* Workers that are started by a provisioner should fetch `TASKCLUSTER_ROOT_URL`
+  from provisioner on start up
+* Workers should require a new config property for taskcluster root URL in
+  their config, and refuse to run if they do not have it and can't fetch it
+  from provisioner configuration (no default allowed)
+* Workers should pass `TASKCLUSTER_ROOT_URL` environment variable to all task
+  processes they create.
+* Workers should pass `TASKCLUSTER_ROOT_URL` to `taskcluster-proxy` when
+  starting it up
+
+## Changes to tasks that use `taskcluster-proxy`
+
+* Tasks that made API calls directly to taskcluster-proxy, and not via a
+  taskcluster client (e.g. using curl directly) should still work, since the
+  proxy knows which taskcluster root url to use from how it was invoked by the
+  worker. 
+
+## Changes to tasks that use a taskcluster-client
+
+If these tasks use an old taskcluster-client, no changes required. Tasks using
+a new taskcluster client should make sure to set the TASKCLSUTER_ROOT_URL based
+on the env var given to them from the worker.
+
+## Changes to `taskcluster-lib-urls`
+
+* Method `ServicesManifest` should be renamed to `APIReferencesManifest`.
+* The following new methods should be added, e.g. for go client:
+
+```go
+func HTTPReference(rootURL string, version string) string
+func AMQPReference(rootURL string, version string) string
```

Comment:
These are equivalent to the existing `ApiReference` and `ExchangesReference` methods, but omit the service.  I'm not sure how that would work?  Also, given that the manifest format allows arbitrary URLs, I think these methods are unnecessary?